File: generate_data.dart
import 'dart:io';

void main() async {
  final mdFile = File('remixed-40d671c6.md');
  final content = await mdFile.readAsString();

  final outDir = Directory('lib/core/constants');

  String escapeDart(String s) {
    return s.replaceAll('\\', '\\\\').replaceAll("'", "\\'");
  }

  List<String> extractNumberedLines(String block) {
    final items = <String>[];
    for (final line in block.trim().split('\n')) {
      final tline = line.trim();
      final m = RegExp(r'^\d+\.\s+(.*)').firstMatch(tline);
      if (m != null) {
        items.add(m.group(1)!.trim());
      }
    }
    return items;
  }

  String sectionBetween(String text, String startMarker, List<String> endMarkers) {
    int idx = text.indexOf(startMarker);
    if (idx == -1) return '';
    idx += startMarker.length;
    int endIdx = text.length;
    for (final em in endMarkers) {
      final i = text.indexOf(em, idx);
      if (i != -1 && i < endIdx) {
        endIdx = i;
      }
    }
    return text.substring(idx, endIdx);
  }

  final categories = [
    ('party_starter', '🎉 PARTY STARTER'),
    ('mild_but_wild', '😜 MILD BUT WILD'),
    ('family_night', '🏠 FAMILY NIGHT'),
    ('spice_it_up', '🔥 SPICE IT UP'),
    ('girls_night', '💅 GIRLS NIGHT OUT'),
    ('guys_unleashed', '💪 GUYS UNLEASHED'),
  ];

  // 1. NHIE
  print('Generating nhie_data.dart ...');
  final nhieEntries = <String>[];
  for (final cat in categories) {
    final catKey = cat.$1;
    final catEmoji = cat.$2;
    final header = '## $catEmoji — NEVER HAVE I EVER';
    final block = sectionBetween(content, header, ['## ', '# ']);
    final items = extractNumberedLines(block);
    for (final item in items) {
      nhieEntries.add("    {'text': '${escapeDart(item)}', 'category': '$catKey'},");
    }
  }
  
  final nhieDart = '''/// Never Have I Ever statement database
class NhieData {
  static const List<Map<String, dynamic>> statements = [
${nhieEntries.join('\n')}
  ];
}
''';
  await File('${outDir.path}/nhie_data.dart').writeAsString(nhieDart);

  // 2. WYR
  print('Generating wyr_data.dart ...');
  final wyrEntries = <String>[];
  for (final cat in categories) {
    final catKey = cat.$1;
    final catEmoji = cat.$2;
    final header = '## $catEmoji — WOULD YOU RATHER';
    final block = sectionBetween(content, header, ['## ', '# ']);
    final items = extractNumberedLines(block);
    for (final item in items) {
      final parts = item.split(RegExp(r'\s+OR\s+', caseSensitive: false));
      if (parts.length == 2) {
        String a = parts[0].replaceAll(RegExp(r'^Would you rather\s+', caseSensitive: false), '');
        String b = parts[1];
        if (b.endsWith('?')) b = b.substring(0, b.length - 1);
        wyrEntries.add("    {'optionA': '${escapeDart(a)}', 'optionB': '${escapeDart(b)}', 'category': '$catKey'},");
      } else {
        wyrEntries.add("    {'optionA': '${escapeDart(item)}', 'optionB': '', 'category': '$catKey'},");
      }
    }
  }

  final wyrDart = '''/// Would You Rather question database
class WyrData {
  static const List<Map<String, dynamic>> questions = [
${wyrEntries.join('\n')}
  ];
}
''';
  await File('${outDir.path}/wyr_data.dart').writeAsString(wyrDart);

  // 3. TRIVIA
  print('Generating trivia_data.dart ...');
  final triviaEntries = <String>[];
  for (final cat in categories) {
    final catKey = cat.$1;
    final catEmoji = cat.$2;
    final header = '## $catEmoji — TRIVIA';
    final block = sectionBetween(content, header, ['## ', '# ']);
    final items = extractNumberedLines(block);
    print('$catKey items: ${items.length}');
    for (final item in items) {
      final m = RegExp(r'Q:\s*(.*?)\s*A:\s*(.*)').firstMatch(item);
      if (m != null) {
        String q = m.group(1)!.trim();
        if (!q.endsWith('?')) q = q + '?';
        q = q.replaceAll('??', '?');
        final a = m.group(2)!.trim();
        triviaEntries.add("    {'q': '${escapeDart(q)}', 'a': '${escapeDart(a)}', 'category': '$catKey'},");
      }
    }
  }

  final triviaDart = '''/// Trivia question database
class TriviaData {
  static const List<Map<String, dynamic>> questions = [
${triviaEntries.join('\n')}
  ];
}
''';
  await File('${outDir.path}/trivia_data.dart').writeAsString(triviaDart);

  // 4. RANDOM
  print('Generating random_data.dart ...');
  final randomHeader = '# 📋 SECTION 5: RANDOM CHALLENGES';
  final randomBlock = sectionBetween(content, randomHeader, ['---']);
  final randomItems = extractNumberedLines(randomBlock);
  final randomLines = randomItems.map((item) => "    '${escapeDart(item)}',").toList();

  final randomDart = '''/// Random challenges database
class RandomData {
  static const List<String> challenges = [
${randomLines.join('\n')}
  ];
}
''';
  await File('${outDir.path}/random_data.dart').writeAsString(randomDart);

  print('Done!');
}

File: generate_data.py
"""
Script to generate all 4 data .dart files from the remixed-40d671c6.md source.
Reads the markdown, parses each section, and writes correct Dart constants.
"""
import re
import os

MD_FILE = os.path.join(os.path.dirname(__file__), 'remixed-40d671c6.md')
OUT_DIR = os.path.join(os.path.dirname(__file__), 'lib', 'core', 'constants')

with open(MD_FILE, 'r', encoding='utf-8') as f:
    content = f.read()

# ─── helpers ────────────────────────────────────────────────────────────
def escape_dart(s):
    """Escape single quotes and backslashes for Dart string literals."""
    return s.replace('\\', '\\\\').replace("'", "\\'")

def extract_numbered_lines(block):
    """Given a text block, extract numbered items like '1. ...' or 'Q: ... A: ...'"""
    items = []
    for line in block.strip().splitlines():
        line = line.strip()
        m = re.match(r'^\d+\.\s+(.*)', line)
        if m:
            items.append(m.group(1).strip())
    return items

def section_between(text, start_marker, end_markers):
    """Extract text between start_marker and the first of end_markers found."""
    idx = text.find(start_marker)
    if idx == -1:
        return ''
    idx += len(start_marker)
    end_idx = len(text)
    for em in end_markers:
        i = text.find(em, idx)
        if i != -1 and i < end_idx:
            end_idx = i
    return text[idx:end_idx]

# ─── category map ───────────────────────────────────────────────────────
CATEGORIES = [
    ('party_starter', '🎉 PARTY STARTER'),
    ('mild_but_wild', '😜 MILD BUT WILD'),
    ('family_night', '🏠 FAMILY NIGHT'),
    ('spice_it_up', '🔥 SPICE IT UP'),
    ('girls_night_out', '💅 GIRLS NIGHT OUT'),
    ('guys_unleashed', '💪 GUYS UNLEASHED'),
]

# ─── 1. NHIE ────────────────────────────────────────────────────────────
print('Generating nhie_data.dart ...')
nhie_entries = []
for cat_key, cat_emoji in CATEGORIES:
    header = f'## {cat_emoji} — NEVER HAVE I EVER'
    block = section_between(content, header, ['## ', '# '])
    items = extract_numbered_lines(block)
    if not items:
        print(f'  WARNING: no NHIE items for {cat_key}')
    for item in items:
        nhie_entries.append(f"    {{'text': '{escape_dart(item)}', 'category': '{cat_key}'}},")
    print(f'  {cat_key}: {len(items)} items')

nhie_dart = f"""/// Never Have I Ever statement database
class NhieData {{
  static const List<Map<String, dynamic>> statements = [
    // -- PARTY STARTER --
{chr(10).join([e for e in nhie_entries if "'category': 'party_starter'" in e])}
    // -- MILD BUT WILD --
{chr(10).join([e for e in nhie_entries if "'category': 'mild_but_wild'" in e])}
    // -- FAMILY NIGHT --
{chr(10).join([e for e in nhie_entries if "'category': 'family_night'" in e])}
    // -- SPICE IT UP --
{chr(10).join([e for e in nhie_entries if "'category': 'spice_it_up'" in e])}
    // -- GIRLS NIGHT OUT --
{chr(10).join([e for e in nhie_entries if "'category': 'girls_night_out'" in e])}
    // -- GUYS UNLEASHED --
{chr(10).join([e for e in nhie_entries if "'category': 'guys_unleashed'" in e])}
  ];
}}
"""
with open(os.path.join(OUT_DIR, 'nhie_data.dart'), 'w', encoding='utf-8') as f:
    f.write(nhie_dart)
print(f'  Total NHIE entries: {len(nhie_entries)}')

# ─── 2. WYR ─────────────────────────────────────────────────────────────
print('\nGenerating wyr_data.dart ...')
wyr_entries = []
for cat_key, cat_emoji in CATEGORIES:
    header = f'## {cat_emoji} — WOULD YOU RATHER'
    block = section_between(content, header, ['## ', '# '])
    items = extract_numbered_lines(block)
    if not items:
        print(f'  WARNING: no WYR items for {cat_key}')
    for item in items:
        # Split on " OR " (case insensitive)
        parts = re.split(r'\s+OR\s+', item, maxsplit=1)
        if len(parts) == 2:
            # Remove leading "Would you rather " from optionA
            a = parts[0]
            a = re.sub(r'^Would you rather\s+', '', a, flags=re.IGNORECASE)
            b = parts[1].rstrip('?')
            wyr_entries.append(f"    {{'optionA': '{escape_dart(a)}', 'optionB': '{escape_dart(b)}', 'category': '{cat_key}'}},")
        else:
            # Fallback: store as single option
            wyr_entries.append(f"    {{'optionA': '{escape_dart(item)}', 'optionB': '', 'category': '{cat_key}'}},")
    print(f'  {cat_key}: {len(items)} items')

wyr_dart = f"""/// Would You Rather question database
class WyrData {{
  static const List<Map<String, dynamic>> questions = [
    // -- PARTY STARTER --
{chr(10).join([e for e in wyr_entries if "'category': 'party_starter'" in e])}
    // -- MILD BUT WILD --
{chr(10).join([e for e in wyr_entries if "'category': 'mild_but_wild'" in e])}
    // -- FAMILY NIGHT --
{chr(10).join([e for e in wyr_entries if "'category': 'family_night'" in e])}
    // -- SPICE IT UP --
{chr(10).join([e for e in wyr_entries if "'category': 'spice_it_up'" in e])}
    // -- GIRLS NIGHT OUT --
{chr(10).join([e for e in wyr_entries if "'category': 'girls_night_out'" in e])}
    // -- GUYS UNLEASHED --
{chr(10).join([e for e in wyr_entries if "'category': 'guys_unleashed'" in e])}
  ];
}}
"""
with open(os.path.join(OUT_DIR, 'wyr_data.dart'), 'w', encoding='utf-8') as f:
    f.write(wyr_dart)
print(f'  Total WYR entries: {len(wyr_entries)}')

# ─── 3. TRIVIA ──────────────────────────────────────────────────────────
print('\nGenerating trivia_data.dart ...')
trivia_entries = []
for cat_key, cat_emoji in CATEGORIES:
    header = f'## {cat_emoji} — TRIVIA'
    block = section_between(content, header, ['## ', '# '])
    items = extract_numbered_lines(block)
    if not items:
        print(f'  WARNING: no TRIVIA items for {cat_key}')
    for item in items:
        # Parse "Q: ... A: ..."
        m = re.match(r'Q:\s*(.*?)\s*A:\s*(.*)', item)
        if m:
            q = m.group(1).strip().rstrip('?') + '?'
            q = q.replace('??', '?')
            a = m.group(2).strip()
            trivia_entries.append(f"    {{'q': '{escape_dart(q)}', 'a': '{escape_dart(a)}', 'category': '{cat_key}'}},")
        else:
            print(f'  WARNING: could not parse trivia: {item[:60]}')
    print(f'  {cat_key}: {len(items)} items')

trivia_dart = f"""/// Trivia question database
class TriviaData {{
  static const List<Map<String, String>> questions = [
    // -- PARTY STARTER --
{chr(10).join([e for e in trivia_entries if "'category': 'party_starter'" in e])}
    // -- MILD BUT WILD --
{chr(10).join([e for e in trivia_entries if "'category': 'mild_but_wild'" in e])}
    // -- FAMILY NIGHT --
{chr(10).join([e for e in trivia_entries if "'category': 'family_night'" in e])}
    // -- SPICE IT UP --
{chr(10).join([e for e in trivia_entries if "'category': 'spice_it_up'" in e])}
    // -- GIRLS NIGHT OUT --
{chr(10).join([e for e in trivia_entries if "'category': 'girls_night_out'" in e])}
    // -- GUYS UNLEASHED --
{chr(10).join([e for e in trivia_entries if "'category': 'guys_unleashed'" in e])}
  ];
}}
"""
with open(os.path.join(OUT_DIR, 'trivia_data.dart'), 'w', encoding='utf-8') as f:
    f.write(trivia_dart)
print(f'  Total TRIVIA entries: {len(trivia_entries)}')

# ─── 4. RANDOM ──────────────────────────────────────────────────────────
print('\nGenerating random_data.dart ...')
header = '# 📋 SECTION 5: RANDOM CHALLENGES'
block = section_between(content, header, ['# ', '---\n\n*End'])
items = extract_numbered_lines(block)
print(f'  Random challenges: {len(items)} items')

random_lines = []
for item in items:
    random_lines.append(f'    "{escape_dart(item)}",')

random_dart = f"""class RandomData {{
  static const List<String> challenges = [
{chr(10).join(random_lines)}
  ];
}}
"""
with open(os.path.join(OUT_DIR, 'random_data.dart'), 'w', encoding='utf-8') as f:
    f.write(random_dart)

print('\n✅ All 4 data files generated successfully!')
